src/frequenz/sdk/timeseries/formulas/_formula_evaluator.py

```diff
@@ -0,0 +1,129 @@
+            root: The root node of the formula AST.
+            components: The telemetry streams that the formula depends on.
+            create_method: A method to generate the output values with.  If the
+                formula is for generating power values, this would be
+                `Power.from_watts`, for example.
+            output_channel: The channel to send evaluated samples to.
+            metric_fetcher: An optional metric fetcher that needs to be started
+                before the formula can be evaluated.
+        """
+        super().__init__()
+
+        self._root: _ast.Node = root
+        self._components: list[_ast.TelemetryStream[QuantityT]] = components
+        self._create_method: Callable[[float], QuantityT] = create_method
+        self._metric_fetcher: ResampledStreamFetcher | None = metric_fetcher
+        self._output_channel: Broadcast[Sample[QuantityT]] = output_channel
+
+        self._output_sender: Sender[Sample[QuantityT]] = output_channel.new_sender()
+
+    @override
+    async def _run(self) -> None:
+        """Run the formula evaluator actor."""
+        if self._metric_fetcher is not None:
+            await self._metric_fetcher.subscribe()
+        await synchronize_receivers(self._components)
+
+        while True:
+            try:
+                timestamp = next(
+                    comp.latest_sample.timestamp
+                    for comp in self._components
+                    if comp.latest_sample is not None
+                )
+
+                res = self._root.evaluate()
+                next_sample = Sample(
+                    timestamp, None if res is None else self._create_method(res)
+                )
+                await self._output_sender.send(next_sample)
+            except (StopAsyncIteration, StopIteration):
+                _logger.debug(
+                    "No more input samples available; stopping formula evaluator. (%s)",
+                    self._root,
+                )
+                await self._output_channel.close()
+                return
+            except Exception as e:  # pylint: disable=broad-except
+                _logger.error(
+                    "Error evaluating formula %s: %s", self._root, e, exc_info=True
+                )
+                await self._output_channel.close()
+                return
+
+            fetch_results = await asyncio.gather(
+                *(comp.fetch_next() for comp in self._components),
+                return_exceptions=True,
+            )
+            if ex := next((e for e in fetch_results if isinstance(e, Exception)), None):
+                if isinstance(ex, (StopAsyncIteration, ReceiverStoppedError)):
+                    _logger.debug(
+                        "input streams closed; stopping formula evaluator. (%s)",
+                        self._root,
+                    )
+                    await self._output_channel.close()
+                    return
+                raise ex
```

Comment:
I think you can potentially simplify this exception handling if you use a `TaskGroup`, something like:


```suggestion
            try:
                async with TaskGroup() as tg:
                    for comp in self._components:
                        tg.create_task(comp.fetch_next())
            except* (StopAsyncIteration, ReceiverStoppedError) as exc:
                _logger.debug(
                    "input streams closed; stopping formula evaluator. (%s): %s",
                    self._root,
                    exc,
                )
                await self._output_channel.close()
                # no return here, otherwise it will interrupt the handling of other exceptions
            if self._output_channel.is_closed:
                return
```

This should create one task per component to fetch and wait until all finished if there are no errors.

If there are errors, all other pending tasks will be cancelled **immediately** (gather will wait until all awaitables finish then `return_exceptions=True`).

If all errors are stop errors, it will log and exit. If there are any other errors, the exception group will bubble up (be re-raised).

Reply:
This stuff has gone away from here in the next PR.  This might still be useful elsewhere, so I'll keep in mind.